File: pycore/pyutils/rpc_v2/common/request_manager.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Request manager for rpc_v2.
"""

import time
import uuid
from typing import Any, Callable, Dict, List, Optional

from pycore.pyfoundations.pybasecommon.color_print import ColorPrint
from pycore.pyutils.rpc_v2.constants import (
    DEFAULT_ACK_MAX_RETRIES,
    DEFAULT_ACK_RETRY_INTERVAL,
    REQUEST_MANAGER_MAX_SIZE,
)
from pycore.pyfoundations.serialized_worker import (
    init_serialized_owner,
    serialized_method,
)


class RequestManager:
    def __init__(self):
        self.requests: Dict[str, Dict[str, Any]] = {}
        self.callbacks: Dict[str, Dict[str, Any]] = {}
        init_serialized_owner(
            self,
            'pyutils.rpc_v2.request_manager',
            'RPCRequestManagerThread',
        )

    @serialized_method
    def create_request(self, session_id: str, metadata: Optional[Dict[str, Any]] = None) -> str:
        request_id = str(uuid.uuid4())
        self.requests[request_id] = {
            "id": request_id,
            "session_id": session_id,
            "created_at": time.time(),
            "status": "pending",
            "retries": 0,
            "max_retries": DEFAULT_ACK_MAX_RETRIES,
            "retry_interval": DEFAULT_ACK_RETRY_INTERVAL,
            "metadata": metadata or {},
        }
        return request_id

    @serialized_method
    def get_request(self, request_id: str) -> Optional[Dict[str, Any]]:
        request = self.requests.get(request_id)
        return dict(request) if request else None

    @serialized_method
    def update_request_status(self, request_id: str, status: str):
        request = self.requests.get(request_id)
        if request:
            request["status"] = status
            request["updated_at"] = time.time()

    @serialized_method
    def increment_retry(self, request_id: str) -> int:
        request = self.requests.get(request_id)
        if request:
            request["retries"] += 1
            return request["retries"]
        return 0

    @serialized_method
    def can_retry(self, request_id: str) -> bool:
        request = self.requests.get(request_id)
        return bool(request and request["retries"] < request["max_retries"])

    @serialized_method
    def register_callback(self, request_id: str, callback: Callable, context: Optional[Any] = None) -> bool:
        if not callable(callback):
            ColorPrint.red("[RequestManager] Callback must be callable")
            return False
        self.callbacks[request_id] = {"callback": callback, "context": context, "created_at": time.time()}
        return True

    @serialized_method
    def _get_callback_info(self, request_id: str) -> Optional[Dict[str, Any]]:
        """Return a detached callback record."""
        callback_info = self.callbacks.get(request_id)
        return dict(callback_info) if callback_info else None

    @serialized_method
    def _finish_callback(self, request_id: str) -> None:
        """Finalize one callback on the state-owner thread."""
        self.update_request_status(request_id, "completed")
        self.callbacks.pop(request_id, None)

    async def execute_callback(self, request_id: str, data: Any, error: Optional[Exception] = None) -> bool:
        callback_info = self._get_callback_info(request_id)
        if not callback_info:
            ColorPrint.yellow(f"[RequestManager] No callback found for request {request_id}")
            return False

        callback = callback_info["callback"]
        context = callback_info.get("context")
        try:
            if context:
                if error:
                    await callback(context, data, error)
                else:
                    await callback(context, data)
            else:
                if error:
                    await callback(data, error)
                else:
                    await callback(data)

            self._finish_callback(request_id)
            return True
        except Exception as exc:
            ColorPrint.red(f"[RequestManager] Callback execution error for {request_id}: {exc}")
            return False

    @serialized_method
    def remove_request(self, request_id: str):
        self.requests.pop(request_id, None)
        self.callbacks.pop(request_id, None)

    @serialized_method
    def get_requests_by_session(self, session_id: str) -> List[str]:
        return [rid for rid, req in self.requests.items() if req["session_id"] == session_id]

    @serialized_method
    def cleanup(self, max_age: float = 3600.0) -> int:
        now = time.time()
        expired = [
            rid
            for rid, req in self.requests.items()
            if now - req.get("created_at", now) > max_age and req.get("status") != "completed"
        ]
        for request_id in expired:
            self.remove_request(request_id)
        return len(expired)


default_request_manager = RequestManager()

__all__ = ["RequestManager", "default_request_manager"]

File: ncore/utils/rpc/common/request_manager.js
const { v4: uuidv4 } = require('uuid');
const logger = require('#@logger');

class RequestManager {
    constructor() {
        this.requests = new Map();
        this.callbacks = new Map();
    }

    createRequest(sessionId, metadata = {}) {
        const requestId = uuidv4();

        this.requests.set(requestId, {
            id: requestId,
            sessionId,
            createdAt: Date.now(),
            status: 'pending',
            retries: 0,
            maxRetries: 3,
            retryInterval: 1000,
            metadata
        });

        return requestId;
    }

    getRequest(requestId) {
        return this.requests.get(requestId);
    }

    hasRequest(requestId) {
        return this.requests.has(requestId);
    }

    updateRequestStatus(requestId, status) {
        const request = this.requests.get(requestId);
        if (request) {
            request.status = status;
            request.updatedAt = Date.now();
        }
    }

    incrementRetry(requestId) {
        const request = this.requests.get(requestId);
        if (request) {
            request.retries++;
            return request.retries;
        }
        return 0;
    }

    canRetry(requestId) {
        const request = this.requests.get(requestId);
        if (request) {
            return request.retries < request.maxRetries;
        }
        return false;
    }

    registerCallback(requestId, callback, context = null) {
        if (typeof callback !== 'function') {
            logger.error('RequestManager: Callback must be a function');
            return false;
        }

        this.callbacks.set(requestId, {
            callback,
            context,
            createdAt: Date.now()
        });

        return true;
    }

    async executeCallback(requestId, data, error = null) {
        const callbackInfo = this.callbacks.get(requestId);

        if (!callbackInfo) {
            logger.warn(`RequestManager: No callback found for request ${requestId}`);
            return false;
        }

        try {
            if (callbackInfo.context) {
                await callbackInfo.callback.call(callbackInfo.context, data, error);
            } else {
                await callbackInfo.callback(data, error);
            }

            this.updateRequestStatus(requestId, 'completed');
            this.callbacks.delete(requestId);

            return true;
        } catch (e) {
            logger.error(`RequestManager: Callback execution error for ${requestId}:`, e);
            return false;
        }
    }

    removeRequest(requestId) {
        this.requests.delete(requestId);
        this.callbacks.delete(requestId);
    }

    getRequestsBySession(sessionId) {
        const requests = [];
        for (const [requestId, request] of this.requests.entries()) {
            if (request.sessionId === sessionId) {
                requests.push(requestId);
            }
        }
        return requests;
    }

    removeRequestsBySession(sessionId) {
        const requests = this.getRequestsBySession(sessionId);
        requests.forEach(requestId => {
            this.removeRequest(requestId);
        });
        return requests.length;
    }

    cleanup(maxAge = 1800000) {
        const now = Date.now();
        let cleaned = 0;

        for (const [requestId, request] of this.requests.entries()) {
            if (now - request.createdAt > maxAge) {
                this.removeRequest(requestId);
                cleaned++;
            }
        }

        if (cleaned > 0) {
            logger.debug(`RequestManager: Cleaned ${cleaned} expired requests`);
        }

        return cleaned;
    }

    getStats() {
        const stats = {
            total: this.requests.size,
            pending: 0,
            completed: 0,
            failed: 0,
            callbacks: this.callbacks.size
        };

        for (const request of this.requests.values()) {
            if (request.status === 'pending') stats.pending++;
            else if (request.status === 'completed') stats.completed++;
            else if (request.status === 'failed') stats.failed++;
        }

        return stats;
    }
}

const defaultRequestManager = new RequestManager();

module.exports = {
    RequestManager,
    defaultRequestManager
};
